---
title: "week 7  eQTL Analysis"
author: "Rahul ladhani"
date: "2023-05-26"
output:
  pdf_document: default
  word_document: default
  html_document: default
---




```{r}
setwd("C:/Users/rahul/OneDrive - University of West London/Desktop/R learning/Genome analysis/week 6  eQTL Analysis")

```



```{r}
#1. Test all gene-SNP pairs and plot a histogram of all p-values
# source("Matrix_eQTL_R/Matrix_eQTL_engine.r");
install.packages("MatrixEQTL")
library(MatrixEQTL)

```


```{r}
#Location of the package with the data files.
base.dir = find.package('MatrixEQTL');
```


```{r}
#Settings
#a) Linear model to use, modelANOVA, modelLINEAR, or modelLINEAR_CROSS
useModel = modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS

```


```{r}
#b) Genotype file name
SNP_file_name = paste(base.dir, "/data/SNP.txt", sep="");
```


```{r}
#c) Gene expression file name
expression_file_name = paste(base.dir, "/data/GE.txt", sep="");
```


```{r}
#d) Covariates file name
# Set to character() for no covariates
covariates_file_name = paste(base.dir, "/data/Covariates.txt", sep="");
```


```{r}
##e) Output file name
output_file_name = tempfile();
```


```{r}
##f) Only associations significant at this level will be saved
pvOutputThreshold = 1e-2;
```

```{r}
##g) Error covariance matrix
# Set to numeric() for identity.
errorCovariance = numeric();
# errorCovariance = read.table("Sample_Data/errorCovariance.txt");
```


```{r}
#h) Load genotype data
snps = SlicedData$new();
snps$fileDelimiter = "\t"; # the TAB character
snps$fileOmitCharacters = "NA"; # denote missing values;
snps$fileSkipRows = 1; # one row of column labels
snps$fileSkipColumns = 1; # one column of row labels
snps$fileSliceSize = 2000; # read file in slices of 2,000 rows
snps$LoadFile(SNP_file_name);

```


```{r}
#i) Load gene expression data
gene = SlicedData$new();
gene$fileDelimiter = "\t"; # the TAB character
gene$fileOmitCharacters = "NA"; # denote missing values;
gene$fileSkipRows = 1; # one row of column labels
gene$fileSkipColumns = 1; # one column of row labels
gene$fileSliceSize = 2000; # read file in slices of 2,000 rows
gene$LoadFile(expression_file_name);
```

```{r}
#j) Load covariates
cvrt = SlicedData$new();
cvrt$fileDelimiter = "\t"; # the TAB character
cvrt$fileOmitCharacters = "NA"; # denote missing values;
cvrt$fileSkipRows = 1; # one row of column labels
cvrt$fileSkipColumns = 1; # one column of row labels
if(length(covariates_file_name)>0) {
cvrt$LoadFile(covariates_file_name);
}

```

```{r}
#k) Run the analysis
me = Matrix_eQTL_engine(
snps = snps,
gene = gene,
cvrt = cvrt,
output_file_name = output_file_name,
pvOutputThreshold = pvOutputThreshold,
useModel = useModel,
errorCovariance = errorCovariance,
verbose = TRUE,
pvalue.hist = TRUE,
min.pv.by.genesnp = FALSE,
noFDRsaveMemory = FALSE);
```

```{r}
#l) Results:
cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n');

```

```{r}

cat('Detected eQTLs:', '\n');

```
```{r}
show(me$all$eqtls)
```

```{r}
#We can also figure look at the number and type of eQTL
#Number of eQTL
show(me$all$neqtls)

```

```{r}
#View the eQTL in the table
View(me$all$eqtls)
```


```{r}
#m) Plot the histogram of all p-values
plot(me)
```

##Test local and distant gene-SNP pairs separately and plot Q-Q plots of local and distant pvalues

```{r}
#source("Matrix_eQTL_R/Matrix_eQTL_engine.r");
library(MatrixEQTL)
```


```{r}
#a) Location of the package with the data files.
base.dir = find.package('MatrixEQTL');
```

```{r}
#b) Settings Linear model to use, modelANOVA, modelLINEAR, or
modelLINEAR_CROSS
useModel = modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS
```

```{r}
##c) Genotype file name
SNP_file_name = paste(base.dir, "/data/SNP.txt", sep="");
```


```{r}
#d) SNP location file name
snps_location_file_name = paste(base.dir, "/data/snpsloc.txt", sep="");
```


```{r}
#e) Gene expression file name
expression_file_name = paste(base.dir, "/data/GE.txt", sep="");
gene_location_file_name = paste(base.dir, "/data/geneloc.txt", sep="");
```


```{r}
#f) Covariates file name
# Set to character() for no covariates
covariates_file_name = paste(base.dir, "/data/Covariates.txt", sep="");
```


```{r}
#g) Output file name
output_file_name_cis = tempfile();
output_file_name_tra = tempfile();
```


```{r}
#h) Only associations significant at this level will be saved
pvOutputThreshold_cis = 2e-2;
pvOutputThreshold_tra = 1e-2;

```


```{r}
#i) Error covariance matrix
# Set to numeric() for identity.
errorCovariance = numeric();
# errorCovariance = read.table("Sample_Data/errorCovariance.txt");
```

```{r}
#j) Distance for local gene-SNP pairs
cisDist = 1e6;
```


```{r}
#k) Load genotype data
snps = SlicedData$new();
snps$fileDelimiter = "\t"; # the TAB character
snps$fileOmitCharacters = "NA"; # denote missing values;
snps$fileSkipRows = 1; # one row of column labels
snps$fileSkipColumns = 1; # one column of row labels
snps$fileSliceSize = 2000; # read file in slices of 2,000 rows
snps$LoadFile(SNP_file_name);
```

```{r}
#l) Load gene expression data
gene = SlicedData$new();
gene$fileDelimiter = "\t"; # the TAB character
gene$fileOmitCharacters = "NA"; # denote missing values;
gene$fileSkipRows = 1; # one row of column labels
gene$fileSkipColumns = 1; # one column of row labels
gene$fileSliceSize = 2000; # read file in slices of 2,000 rows
gene$LoadFile(expression_file_name);
```

```{r}
#m) Load covariates
cvrt = SlicedData$new();
cvrt$fileDelimiter = "\t"; # the TAB character
cvrt$fileOmitCharacters = "NA"; # denote missing values;
cvrt$fileSkipRows = 1; # one row of column labels
cvrt$fileSkipColumns = 1; # one column of row labels
if(length(covariates_file_name)>0) {
cvrt$LoadFile(covariates_file_name);
}
```

```{r}
#n) Run the analysis
snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors
= FALSE);
genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors
= FALSE);
me = Matrix_eQTL_main(
snps = snps,
gene = gene,
cvrt = cvrt,
output_file_name = output_file_name_tra,
pvOutputThreshold = pvOutputThreshold_tra,
useModel = useModel,
errorCovariance = errorCovariance,
verbose = TRUE,
output_file_name.cis = output_file_name_cis,
pvOutputThreshold.cis = pvOutputThreshold_cis,
snpspos = snpspos,
genepos = genepos,
cisDist = cisDist,
pvalue.hist = "qqplot",
min.pv.by.genesnp = FALSE,
noFDRsaveMemory = FALSE);
```

```{r}
#o) Results:
cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n');
```
```{r}
cat('Detected local eQTLs:', '\n');

```

```{r}
show(me$cis$eqtls)

```

```{r}
cat('Detected distant eQTLs:', '\n');

```

```{r}
show(me$trans$eqtls)
```

```{r}
#p) Plot the Q-Q plot of local and distant p-values
plot(me)
```

```{r}
#3. Create an artificial dataset and plot the histogram and Q-Q plot of all p-values
library('MatrixEQTL')
```


```{r}
#a) Number of samples
n = 100;
```


```{r}
#b) Number of variables
ngs = 2000;
```


```{r}
#c) Common signal in all variables (population stratification)
pop = 0.2 * rnorm(n);
```

```{r}
#d) Data matrices
snps.mat = matrix(rnorm(n*ngs), ncol = ngs) + pop;
gene.mat = matrix(rnorm(n*ngs), ncol = ngs) + pop +
snps.mat*((1:ngs)/ngs)^9/2;
```

```{r}
#e) Data objects for Matrix eQTL engine
snps1 = SlicedData$new( t( snps.mat ) );
gene1 = SlicedData$new( t( gene.mat ) );
cvrt1 = SlicedData$new( );
rm(snps.mat, gene.mat)

```

```{r}
#f) Slice data in blocks of 500 variables
snps1$ResliceCombined(500);
gene1$ResliceCombined(500);
```


```{r}
#Name of temporary output file
filename = tempfile();
```


```{r}
#g) Perform analysis recording information for a histogram
meh = Matrix_eQTL_engine(
snps = snps1,
gene = gene1,
cvrt = cvrt1,
output_file_name = filename,
pvOutputThreshold = 1e-100,
useModel = modelLINEAR,
errorCovariance = numeric(),
verbose = TRUE,
pvalue.hist = 100);
```

```{r}
#h) Histogram
#png(filename = "histogram.png", width = 650, height = 650)
plot(meh, col="grey")
```

## dev.off();
```{r}
#i) Perform the same analysis recording information for a Q-Q plot

meq = Matrix_eQTL_engine(
snps = snps1,
gene = gene1,
cvrt = cvrt1,
output_file_name = filename,
pvOutputThreshold = 1e-6,
useModel = modelLINEAR,
errorCovariance = numeric(),
verbose = TRUE,
pvalue.hist = "qqplot");

```

```{r}
# png(filename = "QQplot.png", width = 650, height = 650)
plot(meq, pch = 16, cex = 0.7)
```

# dev.off();


